File: frontend/src/lib/apiClients/leadExport.ts
/**
 * Audited Lead Queue CSV export: the receipt contract and the two digests the
 * browser computes before it asks for one (audit tables-08, wave 1a).
 *
 * The CSV bytes never leave the browser. What the server owns is the ledger
 * row: `POST /api/leads/export-receipt` writes a LEAD_EXPORT event and the
 * download waits for its answer. The client declares what the file holds;
 * the server recomputes the borrower-id digest from the ids it is sent and
 * refuses a declaration that does not describe them.
 */
export type LeadExportScope = 'selected' | 'loaded';

/** The shape of `planLeadCsvExport`'s result this module needs (no component import). */
export interface LeadExportPlanLike {
  scope: 'selected_rows' | 'loaded_rows';
  rows: ReadonlyArray<{ borrower_id: string }>;
}

export interface LeadExportReceiptRequest {
  scope: LeadExportScope;
  row_count: number;
  /** SHA-256 (hex) of the exact CSV text handed to the download. */
  csv_sha256: string;
  /** Masked borrower ids in file order. */
  borrower_ids: string[];
  /** SHA-256 (hex) of `JSON.stringify(borrower_ids)`; verified server-side. */
  borrower_ids_sha256: string;
  /** The queue's query parameters (the CSV's `# filters=` line); fingerprinted server-side. */
  filters: Record<string, string>;
}

export interface LeadExportReceipt {
  audit_event_id: string;
  event_type: 'LEAD_EXPORT';
  actor: string;
  scope: LeadExportScope;
  row_count: number;
  csv_sha256: string;
  borrower_ids_sha256: string;
  filter_fingerprint: string;
  recorded_at: string;
}

const FILTER_KEY_RE = /^[a-z][a-z0-9_]{0,63}$/;

/** Hex SHA-256 of a UTF-8 string through WebCrypto; throws where it is unavailable. */
export async function sha256Hex(text: string): Promise<string> {
  const subtle = globalThis.crypto?.subtle;
  if (!subtle) {
    throw new Error('This browser cannot hash the export for its audit receipt.');
  }
  const digest = await subtle.digest('SHA-256', new TextEncoder().encode(text));
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, '0')).join('');
}

/** The scope token the ledger records for an export plan. */
export function leadExportScope(plan: Pick<LeadExportPlanLike, 'scope'>): LeadExportScope {
  return plan.scope === 'selected_rows' ? 'selected' : 'loaded';
}

/**
 * The queue's query string (`buildLeadQueueExportFilters`, or `none`) as the
 * flat map the receipt fingerprints. Only well-formed parameter names pass;
 * the first value of a repeated key wins, as the API reads it.
 */
export function leadExportFiltersFromQuery(filters: string | undefined | null): Record<string, string> {
  if (!filters || filters === 'none') return {};
  const out: Record<string, string> = {};
  for (const [key, value] of new URLSearchParams(filters)) {
    if (!FILTER_KEY_RE.test(key) || key in out) continue;
    out[key] = value;
  }
  return out;
}

/** The declaration for one planned export, digests included. */
export async function buildLeadExportDeclaration(
  csv: string,
  plan: LeadExportPlanLike,
  filters: string | undefined | null,
): Promise<LeadExportReceiptRequest> {
  const borrowerIds = plan.rows.map((row) => row.borrower_id);
  const [csvSha256, idsSha256] = await Promise.all([
    sha256Hex(csv),
    sha256Hex(JSON.stringify(borrowerIds)),
  ]);
  return {
    scope: leadExportScope(plan),
    row_count: borrowerIds.length,
    csv_sha256: csvSha256,
    borrower_ids: borrowerIds,
    borrower_ids_sha256: idsSha256,
    filters: leadExportFiltersFromQuery(filters),
  };
}

File: backend/schemas/lead_export.py
"""Wire contract for the audited Lead Queue CSV export receipt.

The CSV bytes are produced in the browser from the ``/api/leads`` payload the
operator already saw (no server-streamed export, an owner decision recorded in
the audit register, tables-08). What the server owns is the LEDGER ENTRY: the
client must obtain a ``LEAD_EXPORT`` receipt before the download starts, and
the receipt carries what an auditor needs to tie a file on disk back to a
decision -- the actor, the filter fingerprint, the exported row count, the
SHA-256 of the file bytes and the SHA-256 of the borrower-id list, the latter
recomputed server-side from the ids the client sends.
"""

from __future__ import annotations

import re
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from backend.schemas.common import validate_public_borrower_id
from backend.schemas.lead_query import MAX_LEAD_LIMIT

SHA256_HEX_PATTERN = re.compile(r"^[0-9a-f]{64}$")
_FILTER_KEY_PATTERN = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
MAX_EXPORT_FILTER_ENTRIES = 64
MAX_EXPORT_FILTER_VALUE_LENGTH = 2048

LeadExportScope = Literal["selected", "loaded"]


def _validate_sha256_hex(value: str, field_name: str) -> str:
    digest = str(value).strip().lower()
    if not SHA256_HEX_PATTERN.fullmatch(digest):
        raise ValueError(f"{field_name} must be a lowercase SHA-256 hex digest")
    return digest


class LeadExportReceiptRequest(BaseModel):
    """What the client declares about the file it is about to download."""

    scope: LeadExportScope
    row_count: int = Field(ge=1, le=MAX_LEAD_LIMIT)
    csv_sha256: str = Field(
        description="SHA-256 (hex) of the exact CSV bytes handed to the download.",
    )
    borrower_ids: list[str] = Field(
        min_length=1,
        max_length=MAX_LEAD_LIMIT,
        description="Masked borrower ids in the order the file holds them.",
    )
    borrower_ids_sha256: str = Field(
        description=(
            "SHA-256 (hex) of the compact JSON array of borrower_ids, computed "
            "client-side; the server recomputes it from borrower_ids and refuses "
            "the receipt on a mismatch."
        ),
    )
    filters: dict[str, str] = Field(
        default_factory=dict,
        description=(
            "The Lead Queue query parameters the exported rows were read with "
            "(the same string the CSV's `# filters=` metadata line carries). "
            "Only their fingerprint is written to the ledger."
        ),
    )

    @field_validator("csv_sha256")
    @classmethod
    def _csv_digest(cls, value: str) -> str:
        return _validate_sha256_hex(value, "csv_sha256")

    @field_validator("borrower_ids_sha256")
    @classmethod
    def _ids_digest(cls, value: str) -> str:
        return _validate_sha256_hex(value, "borrower_ids_sha256")

    @field_validator("borrower_ids")
    @classmethod
    def _public_borrower_ids(cls, value: list[str]) -> list[str]:
        normalized = [validate_public_borrower_id(item) for item in value]
        if len(set(normalized)) != len(normalized):
            raise ValueError("borrower_ids must not repeat")
        return normalized

    @field_validator("filters")
    @classmethod
    def _bounded_filters(cls, value: dict[str, str]) -> dict[str, str]:
        if len(value) > MAX_EXPORT_FILTER_ENTRIES:
            raise ValueError("filters carries too many entries")
        for key, raw in value.items():
            if not _FILTER_KEY_PATTERN.fullmatch(key):
                raise ValueError("filters keys must be lowercase query parameter names")
            if not isinstance(raw, str) or len(raw) > MAX_EXPORT_FILTER_VALUE_LENGTH:
                raise ValueError("filters values must be bounded strings")
        return value


class LeadExportReceipt(BaseModel):
    """The ledger entry the download waits for."""

    audit_event_id: str
    event_type: Literal["LEAD_EXPORT"] = "LEAD_EXPORT"
    actor: str
    scope: LeadExportScope
    row_count: int
    csv_sha256: str
    borrower_ids_sha256: str
    filter_fingerprint: str
    recorded_at: str
